src/replication.c

```diff
@@ -1489,17 +1536,30 @@ static int takeSnapshot(struct raft *r)
         goto abort;
     }
 
-    assert(r->snapshot.put.data == NULL);
-    r->snapshot.put.data = r;
-    rv = r->io->snapshot_put(r->io, r->snapshot.trailing, &r->snapshot.put,
-                             snapshot, takeSnapshotCb);
-    if (rv != 0) {
-        goto abort_after_fsm_snapshot;
+    bool sync_snapshot = r->fsm->version < 3 || r->fsm->snapshot_async == NULL;
+    if (sync_snapshot) {
+        /* putSnapshot will clean up config and buffers case of error */
+        return putSnapshot(r, snapshot, takeSnapshotCb);
+    } else {
+        struct raft_io_async_work *take = raft_malloc(sizeof(*take));
+        if (take == NULL) {
+            rv = RAFT_NOMEM;
+            goto abort_after_snapshot;
+        }
+        take->cb = NULL;
```

Comment:
Nit: I'd avoid setting the `cb` field since it's essentially a "private" field reserved for the `raft_io` implementation. The "user" fields are really just `data` and `work`. Not setting it also makes code a tad more clear since you're passing a `takeSnapshotDoneCb` below, and the two callbacks should not be confused.